crypto/bindings/srs.ts

```
@@ -0,0 +1,215 @@
+  // note: these functions are properly typed, thanks to TS template literal types
+  let createSrs = wasm[`caml_${f}_srs_create_parallel`];
+  let getSrs = wasm[`caml_${f}_srs_get`];
+  let setSrs = wasm[`caml_${f}_srs_set`];
+
+  let maybeLagrangeCommitment = wasm[`caml_${f}_srs_maybe_lagrange_commitment`];
+  let lagrangeCommitment = wasm[`caml_${f}_srs_lagrange_commitment`];
+  let setLagrangeBasis = wasm[`caml_${f}_srs_set_lagrange_basis`];
+  let getLagrangeBasis = wasm[`caml_${f}_srs_get_lagrange_basis`];
+
+  return {
+    /**
+     * returns existing stored SRS or falls back to creating a new one
+     */
+    create(size: number): WasmSrs {
+      let srs = srsStore[f][size] as WasmSrs | undefined;
+
+      if (srs === undefined) {
+        if (cache === undefined) {
+          // if there is no cache, create SRS in memory
+          srs = createSrs(size);
+        } else {
+          let header = cacheHeaderSrs(f, size);
+
+          // try to read SRS from cache / recompute and write if not found
+          try {
+            let bytes = cache.read(header);
+
+            // TODO: this takes a bit to long, about 300ms for 2^16
+            // `pointsToRust` is the clear bottleneck
+            let jsonSrs: OrInfinityJson[] = JSON.parse(
+              new TextDecoder().decode(bytes)
+            );
+            let mlSrs = MlArray.mapTo(jsonSrs, OrInfinity.fromJSON);
+            let wasmSrs = conversion[f].pointsToRust(mlSrs);
+
+            srs = setSrs(wasmSrs);
+          } catch (e) {
+            // not in cache
+            srs = createSrs(size);
+
+            if (cache.canWrite) {
+              let wasmSrs = getSrs(srs);
+
+              let mlSrs = conversion[f].pointsFromRust(wasmSrs);
+              let jsonSrs = MlArray.mapFrom(mlSrs, OrInfinity.toJSON);
+              let bytes = new TextEncoder().encode(JSON.stringify(jsonSrs));
+
+              try {
+                cache.write(header, bytes);
+              } catch (e) {}
```

we don't log the error?

---

I wrote this before the cache had a `canWrite` property. so I had to swallow the error here to allow dummy / non-writing caches without spamming the console. However, now that there is `canWrite`, I will actually log the error here